pkg/segment/aggregations/segaggs.go

```diff
@@ -2408,6 +2414,665 @@ func performValueColRequest(nodeResult *structs.NodeResult, aggs *structs.QueryA
+		}
+	}
+
+	if len(nodeResult.Histogram) > 0 {
+		err := performBinRequestOnHistogram(nodeResult, letColReq)
+		if err != nil {
+			return fmt.Errorf("performBinRequest: Error while performing bin request on histogram, err: %v", err)
+		}
+	}
+
+	if len(nodeResult.MeasureResults) > 0 {
+		err := performBinRequestOnMeasureResults(nodeResult, letColReq)
+		if err != nil {
+			return fmt.Errorf("performBinRequest: Error while performing bin request on measure results, err: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func performBin(value float64, spanOptions *structs.BinSpanOptions, binReq *structs.BinCmdOptions) (interface{}, error) {
+	if spanOptions != nil {
+		if binReq.Field == "timestamp" {
+			return performBinWithSpanTime(value, spanOptions, binReq.AlignTime)
+		}
+		return performBinWithSpan(value, spanOptions)
+	}
+
+	return nil, fmt.Errorf("performBin: BinSpanOptions is nil")
+}
+
+// This function either returns a float or a string
+func performBinWithSpan(value float64, spanOpt *structs.BinSpanOptions) (interface{}, error) {
+	if spanOpt.BinSpanLength != nil {
+		lowerBound, upperBound := getBinRange(value, spanOpt.BinSpanLength.Num)
+		if spanOpt.BinSpanLength.TimeScale == 0 {
+			return fmt.Sprintf("%v-%v", lowerBound, upperBound), nil
+		} else {
+			return lowerBound, nil
+		}
+	}
+
+	if spanOpt.LogSpan != nil {
+		val := value / spanOpt.LogSpan.Coefficient
+		logVal := math.Log10(val) / math.Log10(spanOpt.LogSpan.Base)
+		floorVal := math.Floor(logVal)
+		ceilVal := math.Ceil(logVal)
+		if ceilVal == floorVal {
+			ceilVal += 1
+		}
+
+		lowerBound := math.Pow(spanOpt.LogSpan.Base, floorVal) * spanOpt.LogSpan.Coefficient
+		upperBound := math.Pow(spanOpt.LogSpan.Base, ceilVal) * spanOpt.LogSpan.Coefficient
+
+		return fmt.Sprintf("%v-%v", lowerBound, upperBound), nil
+	}
+
+	return "", fmt.Errorf("performBinWithSpan: BinSpanLength is nil")
+}
+
+func getTimeBucketWithAlign(utcTime time.Time, durationScale time.Duration, spanOpt *structs.BinSpanOptions, alignTime *uint64) int {
+	if alignTime == nil {
+		return int(utcTime.Truncate(time.Duration(spanOpt.BinSpanLength.Num) * durationScale).UnixMilli())
+	}
+
+	factorInMillisecond := float64((time.Duration(spanOpt.BinSpanLength.Num) * durationScale) / time.Millisecond)
+	currTime := float64(utcTime.UnixMilli())
+	baseTime := float64(*alignTime)
+	diff := math.Floor((currTime - baseTime) / factorInMillisecond)
+	bucket := int(baseTime + diff*factorInMillisecond)
+	if bucket < 0 {
+		bucket = 0
+	}
+
+	return bucket
+}
+
+func findBucketMonthQuarter(utcTime time.Time, num int) uint64 {
+	var finalTime time.Time
+	if num == 12 {
+		finalTime = time.Date(utcTime.Year(), 1, 1, 0, 0, 0, 0, time.UTC)
+	} else {
+		currMonth := int(utcTime.Month())
+		month := ((currMonth-1)/num)*num + 1
+		finalTime = time.Date(utcTime.Year(), time.Month(month), 1, 0, 0, 0, 0, time.UTC)
+	}
+
+	return uint64(finalTime.UnixMilli())
+}
```

Comment:
This is a bit confusing. You should rename `num` to something more descriptive. I think it's the month, but it's not clear, and I'm not sure why you also need the UTC time. Also for `num` it's not clear what the range of valid inputs is (e.g. 1-12, or maybe something else)

Reply:
addressed in the latest commit.